File: Java/chapter_11_challenge/app/src/main/java/com/glella/criminalintent/CrimePagerActivity.java
package com.glella.criminalintent;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageButton;

import java.util.List;
import java.util.UUID;

public class CrimePagerActivity extends AppCompatActivity {

    private static final String EXTRA_CRIME_ID = "com.gella.crime_id";

    private ViewPager mViewPager;
    private List<Crime> mCrimes;

    //Challenge
    private ImageButton mFirstButton;
    private ImageButton mLastButton;

    public static Intent newIntent(Context packageContext, UUID crimeID) {
        Intent intent = new Intent(packageContext, CrimePagerActivity.class);
        intent.putExtra(EXTRA_CRIME_ID, crimeID);
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_crime_pager);

        UUID crimeID = (UUID) getIntent().getSerializableExtra(EXTRA_CRIME_ID);

        mViewPager = (ViewPager) findViewById(R.id.crime_view_pager);

        // Challenge
        mFirstButton = (ImageButton) findViewById(R.id.first_button);
        mLastButton = (ImageButton) findViewById(R.id.last_button);

        mFirstButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mViewPager.setCurrentItem(0); // First Item
            }
        });

        mLastButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mViewPager.setCurrentItem(mCrimes.size()-1); // Last Item
            }
        });

        mCrimes = CrimeLab.get(this).getCrimes();

        FragmentManager fm = getSupportFragmentManager();
        mViewPager.setAdapter(new FragmentStatePagerAdapter(fm) {
            @Override
            public Fragment getItem(int position) {
                Crime crime = mCrimes.get(position);

                // Challenge - enabling and disabling buttons // rethink not working
                if (position == 0) {
                    mFirstButton.setEnabled(false);
                } else {
                    mFirstButton.setEnabled(true);
                }

                if (position == mCrimes.size()-1) {
                    mLastButton.setEnabled(false);
                } else {
                    mLastButton.setEnabled(true);
                }

                return CrimeFragment.newInstance(crime.getId());
            }

            @Override
            public int getCount() {
                return mCrimes.size();
            }
        });

        for (int i = 0; i < mCrimes.size(); i++) {
            if (mCrimes.get(i).getId().equals(crimeID)) {
                mViewPager.setCurrentItem(i);
                break;
            }
        }
    }
}

File: Kotlin/chapter_11_challenge/app/src/main/java/com/glella/criminalintent/CrimePagerActivity.kt
package com.glella.criminalintent

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.ImageButton
import com.glella.criminalintent.CrimeLab.mCrimes
import java.util.*

class CrimePagerActivity: AppCompatActivity() {

    companion object {
        private val EXTRA_CRIME_ID = "com.gella.crime_id"

        fun newIntent(packageContext: Context, crimeID: UUID): Intent {
            val intent = Intent(packageContext, CrimePagerActivity::class.java)
            intent.putExtra(EXTRA_CRIME_ID, crimeID)
            return intent
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_crime_pager)

        val crimeID = intent.getSerializableExtra(EXTRA_CRIME_ID) as UUID

        val mViewPager = findViewById<View>(R.id.crime_view_pager) as ViewPager

        // Challenge
        val mFirstButton = findViewById<View>(R.id.first_button) as ImageButton
        val mLastButton = findViewById<View>(R.id.last_button) as ImageButton

        mFirstButton.setOnClickListener { v ->
            mViewPager.currentItem = 0 // First Item
        }

        mLastButton.setOnClickListener { v ->
            mViewPager.currentItem = mCrimes.size - 1 // Last Item
        }



        mViewPager.adapter = object : FragmentStatePagerAdapter(supportFragmentManager) {
            override fun getItem(position: Int): Fragment {
                val crime = mCrimes.get(position)
                return CrimeFragment.newInstance(crime.mId)
            }

            override fun getCount(): Int {
                return mCrimes.size
            }
        }

        for (i in mCrimes.indices) {
            if (mCrimes[i].mId.equals(crimeID)) {
                mViewPager.currentItem = i
                break
            }
        }


    }
}